### PR TITLE
Factor python dependency out of interpreter

### PR DESCRIPTION
**Pull request by zdevito**
This is part of being able to build our python bindings separately from libtorch.

### CODE REVIEW

**Comment by zdevito** [comment]
@ezyang -- this is similar to the libaten_gpu.so stuff, but for removing the NO_PYTHON ifdef. This is just the intepreter changes. Can you take a look?

**Review by ezyang — APPROVED** [review]
LGTM. It seems like a good idea in general to use the Python initialization functions to initialize Python related state, rather than static initializers; let's make sure other people doing Python separation work keep this in mind.

File: torch/csrc/jit/interpreter.cpp [hunk 1]
```diff
@@ -17,18 +14,27 @@
...
+// can be in a separate compilation unit
+static std::mutex handler_mutex;
+static std::vector<OpHandler> handlers;
+void addInterpreterOpHandler(OpHandler handler) {
+  std::lock_guard<std::mutex> guard(handler_mutex);
+  handlers.push_back(handler);
+}
+at::optional<Operation> lookupExternalOp(Node* n) {
+  std::lock_guard<std::mutex> guard(handler_mutex);
+  for(auto & handler : handlers) {
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: I know you are joking, but the indention here is that each handler doesn't check a single op, but that each handler checks _all_ the ops that a shared library has added. So the number of handlers should be a small constant.

**Pull request merged**